battle_test.go

```diff
@@ -6,63 +6,65 @@ import (
 	"testing"
 )
 
-// Blindly uses the first move on the first opponent pokemon.
 type dumbAgent struct{}
 
-func (dumbAgent) Act(b BattleInfo) Turn {
-	opponent := b.Opponents()[0]
-	return FightTurn{
-		moveIdx:   0,
-		targetIdx: opponent,
+// Blindly uses the first move on the first opponent pokemon.
+func (a dumbAgent) Act(ctx *BattleContext) Turn {
+	// You can use `a` (reference to self) for self-targeting turns
+	for agent, party := range ctx.Opponents {
+		for i := range party {
+			return FightTurn{
+				agent:  agent,
+				move:   0,
+				target: i,
+			}
+		}
 	}
+	panic("no opponents found")
 }
 
 func TestBattleSetup(t *testing.T) {
 	a1 := Agent(dumbAgent{})
 	a2 := Agent(dumbAgent{})
-	b := Battle{}
-	party1 := Party{
-		Agent: &a1,
-	}
-	pkmn1 := GetPokemon(4)
+	party1 := NewParty(&a1, 0)
+	pkmn1 := NewPokemon(4)
 	party1.AddPokemon(&pkmn1)
-	party2 := Party{
-		Agent: &a2,
-	}
-	pkmn2 := GetPokemon(7)
+	party2 := NewParty(&a2, 1)
+	pkmn2 := NewPokemon(7)
 	party2.AddPokemon(&pkmn2)
-	b.AddParty(&party1, &party2)
-	b.SetTeams([][]int{{0}, {1}})
+	b := NewBattle()
+	b.AddParty(party1, party2)
 }
 
 func TestBattleOneRound(t *testing.T) {
 	a1 := Agent(dumbAgent{})
 	a2 := Agent(dumbAgent{})
-	b := Battle{}
-	party1 := Party{
-		Agent: &a1,
-	}
-	pound := GetMove(1)
-	pkmn1 := GetPokemon(4)
+	party1 := NewParty(&a1, 0)
+	pkmn1 := NewPokemon(4)
+	pound := NewMove(1)
 	pkmn1.Moves[0] = &pound
 	party1.AddPokemon(&pkmn1)
-	party2 := Party{
-		Agent: &a2,
-	}
-	pkmn2 := GetPokemon(7)
+	party2 := NewParty(&a2, 1)
+	pkmn2 := NewPokemon(7)
 	pkmn2.Moves[0] = &pound
 	party2.AddPokemon(&pkmn2)
-	b.AddParty(&party1, &party2)
-	b.SetTeams([][]int{{0}, {1}})
-
-	b.Start()
-	for p, party := range b.Parties {
-		got := party.GetActive()
-		if !reflect.DeepEqual(got, []int{0}) {
+	b := NewBattle()
+	b.AddParty(party1, party2)
+	err := b.Start()
+	if err != nil {
+		t.Fatal("failed to start battle")
+	}
+	for p, party := range b.parties {
+		got := party.GetActivePokemon()
+		if !reflect.DeepEqual(got, map[int]Pokemon{0: *party.pokemon[0]}) {
 			t.Fatalf("Must send out first pokemon in each at the beginning of the battle. Party %d gave: %v", p, got)
 		}
 	}
 	b.SimulateRound()
+	party1.SetInactive(0)
+	if len(b.parties[0].activePokemon) != 0 {
+		t.Error("expected party 1 to have no active Pokemon")
+	}
```

Comment:
IMO, this should go in a separate test just to test `SetActive` and `SetInactive`. It also conflicts with the fight turn PR I'm working on.